src/librustc_mir/util/pretty.rs

```diff
@@ -534,6 +538,174 @@ pub fn write_mir_intro<'tcx>(
     Ok(())
 }
 
+/// Find all `AllocId`s mentioned (recursively) in the MIR body and print their corresponding
+/// allocations.
+pub fn write_allocations<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    body: &Body<'_>,
+    w: &mut dyn Write,
+) -> io::Result<()> {
+    fn ids(alloc: &Allocation) -> impl Iterator<Item = AllocId> + '_ {
+        alloc.relocations().iter().map(|(_, (_, id))| *id)
+    }
+    fn cids(val: ConstValue<'_>, mut f: impl FnMut(AllocId)) {
+        match val {
+            ConstValue::Scalar(interpret::Scalar::Ptr(ptr)) => {
+                f(ptr.alloc_id);
+            }
+            ConstValue::Scalar(interpret::Scalar::Raw { .. }) => {}
+            ConstValue::ByRef { alloc, .. } | ConstValue::Slice { data: alloc, .. } => {
+                for id in ids(alloc) {
+                    f(id);
+                }
+            }
+        }
+    }
+    struct V(FxHashSet<AllocId>);
+    impl Visitor<'_> for V {
+        fn visit_constant(&mut self, constant: &Constant<'tcx>, location: Location) {
+            self.super_constant(constant, location);
+            let val = match constant.literal.val {
+                ty::ConstKind::Value(val) => val,
+                _ => return,
+            };
+            cids(val, |id| {
+                self.0.insert(id);
+            });
+        }
+    }
+    let mut visitor = V(Default::default());
+    visitor.visit_body(body);
+    let mut seen = visitor.0;
+    let mut todo: Vec<_> = seen.iter().copied().collect();
+    while let Some(id) = todo.pop() {
+        let mut write_allocation = |w: &mut dyn Write, alloc: &Allocation| -> io::Result<()> {
```

Comment:
Given that there's also a function `write_allocation`, maybe call this closure something else... I was quite confused for a bit trying to figure out why the code below could call that function with the wrong number of arguments.^^

Reply:
XD but writing an allocation is what it does! This is just the more locally adjusted version of the general function. I do this kind of shadowing a lot.

Reply:
The other one only writes the bytes inside the allocation, this one also writes parts of the header. So they don't have entirely the same role either.

> I do this kind of shadowing a lot.

Not sure how that counters the argument of such code being hard to read. ;)